### DISCUSSION TITLE
How do I authenticate users signing in using oauth providers(for e.g. GitHub)?

### DISCUSSION THREAD

**Discussion by Maddoxx88** [supabase/supabase#692] [Questions]
I have already registered my OAuth app on GitHub and also tried using the supabase built-in signIn method but it isn't redirecting me to the OAuth GitHub login page so instead I used a hyperlink but after signing in which token do I use to authenticate that particular user?
Tried this code block:
``const { user, error } = await supabase.auth.signIn({
  // provider can be 'github', 'google', 'gitlab', or 'bitbucket'
  provider: 'github'
})``

**Comment by AndreaBarghigiani** [comment]
Hi @Maddoxx88, 
probably I am not the most experienced person to answer this question but in my application (using React) I have done the following.

First and foremost I created a button that will fire the `signIn` when it get's the `onClick` event like so:
```jsx
<Button
	onClick={() => {
		supabase.auth.signIn({ provider: 'bitbucket' });
	}}
	colorScheme="green"
>
	Login
</Button>
```

Since I am using the Context API I am storing the returning value in it and I use `useEffect` to check if the user is logged in the `session` prop that Supabase populates from the call of `createClient` during initialization, I believe you already have done this so I skip this part but you can [check the docs](https://supabase.io/docs/client/initializing) to get a better sense of what I am talking about.

```jsx
export function UserProvider(props) {
	const [session, setSession] = React.useState(null);
	const [user, setUser] = React.useState(null);
	const [publicUser, setPublicUser] = React.useState(null);

	// Check if we have a user logged in
	React.useEffect(() => {
		const session = supabase.auth.session();
		// console.log('session', session);
		setSession(session);
		setUser(session?.user ?? null);
		if (session?.user) {
			setTimeout(async () => {
				const publicUser = await getPublicUser(session.user);
				setPublicUser(publicUser);
			}, 100);
		}
		const { data: authListener } = supabase.auth.onAuthStateChange(
			async (event, session) => {
				if (session?.user) {
					const publicUser = await getPublicUser(session.user);
					setSession(session);
					setUser(session.user);
					setPublicUser(publicUser);
				}

				// @todo: implement redirect on logout
			}
		);

		// Cleanup function
		return () => {
			authListener.unsubscribe();
		};
	}, []);

	const value = [publicUser, user, session];

	return <UserContext.Provider value={value} {...props} />;
}
```
Inside the `useEffect` I also fire the `onAuthStateChange` ([docs](https://supabase.io/docs/client/auth-onauthstatechange)) that fires any time something in the Auth changes, aka the user is logged in or out.

I am sure my code could be improved thousands times but app is working, people can get in logging with BitBucket (in my case) so for now it is good enough.

As stated I am not that experienced when we talk about authentication but I want to improve so if anyone reading this discussion want to share his two cents will be welcomed 😊

*Side note:* as you can see I am also handling with frontend code the creation/pull of a 'public user', it is just a bunch of information that I want to make available on the FE and didn't want to expose the `private.auth` table.

**Reply by Maddoxx88** [reply]
> Hi @Maddoxx88, 
> probably I am not the most experienced person to answer this question but in my application (using React) I have done the following.
> 
> First and foremost I created a button that will fire the `signIn` when it get's the `onClick` event like so:
> ```jsx
> <Button
> 	onClick={() => {
> 		supabase.auth.signIn({ provider: 'bitbucket' });
> 	}}
> 	colorScheme="green"
> >
> 	Login
> </Button>
> ```
> 
> Since I am using the Context API I am storing the returning value in it and I use `useEffect` to check if the user is logged in the `session` prop that Supabase populates from the call of `createClient` during initialization, I believe you already have done this so I skip this part but you can [check the docs](https://supabase.io/docs/client/initializing) to get a better sense of what I am talking about.
> 
> ```jsx
> export function UserProvider(props) {
> 	const [session, setSession] = React.useState(null);
> 	const [user, setUser] = React.useState(null);
> 	const [publicUser, setPublicUser] = React.useState(null);
> 
> 	// Check if we have a user logged in
> 	React.useEffect(() => {
> 		const session = supabase.auth.session();
> 		// console.log('session', session);
> 		setSession(session);
> 		setUser(session?.user ?? null);
> 		if (session?.user) {
> 			setTimeout(async () => {
> 				const publicUser = await getPublicUser(session.user);
> 				setPublicUser(publicUser);
> 			}, 100);
> 		}
> 		const { data: authListener } = supabase.auth.onAuthStateChange(
> 			async (event, session) => {
> 				if (session?.user) {
> 					const publicUser = await getPublicUser(session.user);
> 					setSession(session);
> 					setUser(session.user);
> 					setPublicUser(publicUser);
> 				}
> 
> 				// @todo: implement redirect on logout
> 			}
> 		);
> 
> 		// Cleanup function
> 		return () => {
> 			authListener.unsubscribe();
> 		};
> 	}, []);
> 
> 	const value = [publicUser, user, session];
> 
> 	return <UserContext.Provider value={value} {...props} />;
> }
> ```
> Inside the `useEffect` I also fire the `onAuthStateChange` ([docs](https://supabase.io/docs/client/auth-onauthstatechange)) that fires any time something in the Auth changes, aka the user is logged in or out.
> 
> I am sure my code could be improved thousands times but app is working, people can get in logging with BitBucket (in my case) so for now it is good enough.
> 
> As stated I am not that experienced when we talk about authentication but I want to improve so if anyone reading this discussion want to share his two cents will be welcomed 😊
> 
> *Side note:* as you can see I am also handling with frontend code the creation/pull of a 'public user', it is just a bunch of information that I want to make available on the FE and didn't want to expose the `private.auth` table.

Dude, your authentication looks quite elaborate. I will try catching the current user and session using useEffect. Yet one thing which will remain is when "Sign in with GitHub" button is clicked I land on the redirect_uri which I've mentioned in the GitHub OAuth Settings. How do I catch the right token from there to probably authenticate it using my own way?
Appreciate the help.

**Reply by AndreaBarghigiani** [reply]
Yep the authentication is quite complex indeed because it's an evolution of the codebase that was created for a previous example in Next.js that's now it's refactored using the [@supabase/ui](https://github.com/supabase/ui) components, specifically the `Auth` one.

You can get away using just [that one](https://github.com/supabase/ui/tree/develop/src/components/Auth) but if you like to know why it is that complex please keep reading.

So the essential part about the population of the `UserContext` is the following:

```jsx
export const UserContextProvider = (props) => {
  const [session, setSession] = useState(null)
  const [user, setUser] = useState(null)

  useEffect(() => {
    const session = supabase.auth.session()
    setSession(session)
    setUser(session?.user ?? null)
    const { data: authListener } = supabase.auth.onAuthStateChange(async (event, session) => {
      setSession(session)
      setUser(session?.user ?? null)
    })

    return () => {
      authListener.unsubscribe()
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const value = { session, user }
  return <UserContext.Provider value={value} {...props} />
}
```

In my app I want to separate the information that user will add, like name, role etc... from the user that is used for authentication, for this you see some additional `publicUser` stuff there 😊

I do not follow you when you say *authenticate it using my own way*, do you mean maybe that you do not want to use the GitHub implementation in Supabase and implement your own? If yes unfortunately I don't know 😃
